### PR TITLE
Before running the SCSI.ReceiveCopyResults.CopyStatus test we should check that it is supported

### PR DESCRIPTION
**Pull request by Ser01x**
If ReceiveCopyResults is not implemented then we should skip this test.

### CODE REVIEW

File: test-tool/test_receive_copy_results_copy_status.c [hunk 1]
```diff
@@ -35,10 +35,22 @@ test_receive_copy_results_copy_status(void)
         int  offset = XCOPY_DESC_OFFSET, list_id = 1;
         struct iscsi_data data;
         unsigned char *xcopybuf;
+        struct scsi_command_descriptor *cd;
 
         logging(LOG_VERBOSE, LOG_BLANK_LINE);
         logging(LOG_VERBOSE, "Test RECEIVE COPY RESULTS, COPY STATUS");
 
+        logging(LOG_VERBOSE, "Check that RECEIVE COPY RESULTS is supported.");
+        cd = get_command_descriptor(SCSI_OPCODE_RECEIVE_COPY_RESULTS,
```
Comment (bvanassche) [inline test-tool/test_receive_copy_results_copy_status.c:44]: Implementing REPORT SUPPORTED OPERATION CODES is not mandatory. E.g. LIO does not implement it. I think we need a better solution to test whether or not RECEIVE COPY RESULTS is supported.

Comment (roolebo) [reply test-tool/test_receive_copy_results_copy_status.c:44]: @bvanassche What do you think if we try RSOC first, then try to send the command and if neither works the tests can be skipped?

BTW, we've got RSOC implemented for TCM just haven't managed to send patches yet.

Comment (roolebo) [reply test-tool/test_receive_copy_results_copy_status.c:44]: Another option could be Third-party Copy VPD (0x8f) page as defined in SPC:
```
/* 5.19.1 General considerations for third-party copies */
Application clients and other copy managers request third-party copy operations by sending commands (see 5.19.3) to a copy manager for processing by that copy manager. Copy managers provide information about their capabilities to application clients in the Third-party Copy VPD page (see 7.7.18).
/* 7.7.18.4.1 Supported Commands third-party copy descriptor overview */
The Supported Commands third-party copy descriptor (see table 587) indicates which commands (i.e., combi- nations of operation code and service action) (see 5.19.3) the copy manager supports.
```

Unfortunately it's not implemented in TCM, but it's something that might help feature discovery or test selection along with RSOC.

Comment (Ser01x) [reply test-tool/test_receive_copy_results_copy_status.c:44]: Yes, implement Third-Party Copy VPD in LIO is a good idea. We will do it.

File: lib/scsi-lowlevel.c [hunk 2]
```diff
@@ -1495,6 +1495,102 @@ scsi_inquiry_unmarshall_device_identification(struct scsi_task *task)
...
+			command->service_action[i] = lptr[2 + i];
+		}
+
+		remaining -= command->service_action_length + 2;
+		lptr += command->service_action_length + 2;
+	}
+	return supported_commands;
+
+err:
+	while (supported_commands->commands_supported) {
```
Comment (bvanassche) [inline lib/scsi-lowlevel.c:1545]: Should a free() call perhaps be inserted in this loop?

**Review by bvanassche — CHANGES_REQUESTED** [review]

File: lib/scsi-lowlevel.c (hunk 2 above)
Comment (Ser01x) [reply lib/scsi-lowlevel.c]: I also checked everything with different conditions to use the `valgrind` and did not find the memory leak. `valgrind` shows two no important leaks. It is 'initiatorname1' and 'initiatorname2' variables in `main`:
```
dev-ssamoylenko:~/src/libiscsi/test-tool # valgrind --leak-check=full --show-leak-kinds=all ./iscsi-test-cu -V --dataloss --test SCSI.ReceiveCopyResults.CopyStatus -i iqn.1996-04.de.suse:01:d083346a433 iscsi://172.20.9.152:3260/iqn.2003-
01.org.linux-iscsi.sp-0.ppc64le:sn.f7433c91aa3e/0
==26677== Memcheck, a memory error detector
==26677== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==26677== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==26677== Command: ./iscsi-test-cu -V --dataloss --test SCSI.ReceiveCopyResults.CopyStatus -i iqn.1996-04.de.suse:01:d083346a433 iscsi://172.20.9.152:3260/iqn.2003-01.org.linux-iscsi.sp-0.ppc64le:sn.f7433c91aa3e/0
==26677==
    Send PRIN/READ_KEYS
    Send READCAPACITY10 (Expecting SUCCESS) LBA:0 pmi:0
    [OK] READCAPACITY10 returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send READCAPACITY16 (Expecting SUCCESS)
    [OK] READCAPACITY16 returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:0 page_code:00 alloc_len:64
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:b0 alloc_len:64
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:b1 alloc_len:255
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send REPORT_SUPPORTED_OPCODE (Expecting SUCCESS) RCTD:1 OPTIONS:0 OPCODE:0x00 SA:0 ALLOC_LEN:65535
    [OK] REPORT_SUPPORTED_OPCODES returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Send MODESENSE6 (Expecting SUCCESS)
    [OK] MODESENSE6 returned SUCCESS NO SENSE(0x00) (null)(0x0000)



     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: ReceiveCopyResults
  Test: CopyStatus ...
    Test RECEIVE COPY RESULTS, COPY STATUS
    Get VPD pages supported list.
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:00 alloc_len:255
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    Check that RECEIVE COPY RESULTS is supported.
    Third-party Copy VPD page is supported.
    Send INQUIRY (Expecting SUCCESS) evpd:1 page_code:8f alloc_len:1024
    [OK] INQUIRY returned SUCCESS NO SENSE(0x00) (null)(0x0000)
    [SKIPPED] RECEIVE COPY RESULTS is not implemented.
passed    Send PRIN/READ_KEYS


Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests      1      1      1      0        0
             asserts      6      6      6      0      n/a

Elapsed time =    0.034 seconds
Tests completed with return value: 0
==26677==
==26677== HEAP SUMMARY:
==26677==     in use at exit: 35 bytes in 1 blocks
==26677==   total heap usage: 125 allocs, 124 frees, 33,599,771 bytes allocated
==26677==
==26677== 35 bytes in 1 blocks are still reachable in loss record 1 of 1
==26677==    at 0x4C2E2DF: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==26677==    by 0x52DE749: strdup (in /lib64/libc-2.26.so)
==26677==    by 0x4031CC: main (iscsi-test-cu.c:1178)
==26677==
==26677== LEAK SUMMARY:
==26677==    definitely lost: 0 bytes in 0 blocks
==26677==    indirectly lost: 0 bytes in 0 blocks
==26677==      possibly lost: 0 bytes in 0 blocks
==26677==    still reachable: 35 bytes in 1 blocks
==26677==         suppressed: 0 bytes in 0 blocks
==26677==
==26677== For lists of detected and suppressed errors, rerun with: -s
==26677== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

File: lib/scsi-lowlevel.c [hunk 3]
```diff
@@ -1495,6 +1495,97 @@ scsi_inquiry_unmarshall_device_identification(struct scsi_task *task)
...
+	struct scsi_inquiry_third_party_copy *inq = scsi_malloc(task,
+			sizeof(*inq));
+	int remaining;
+	unsigned char *dptr;
+
+	if (inq == NULL) {
+		return NULL;
+	}
+
+	inq->qualifier             = (task_get_uint8(task, 0) >> 5) & 0x07;
```
Comment (bvanassche) [inline lib/scsi-lowlevel.c:1560]: There are plenty of spaces in front of the assignment operator. Consider using less whitespace.

**Review by bvanassche — CHANGES_REQUESTED** [review]

File: test-tool/test_receive_copy_results_copy_status.c [hunk 4]
```diff
@@ -26,19 +27,112 @@
 #include "scsi-lowlevel.h"
 #include "iscsi-test-cu.h"
 
+
+
```
Comment (bvanassche) [inline test-tool/test_receive_copy_results_copy_status.c:31]: Is it necessary to insert blank lines here? Isn't a single blank line between #include directives and code enough?

File: lib/scsi-lowlevel.c (hunk 2 above)
Comment (bvanassche) [reply lib/scsi-lowlevel.c]: Calling free() for memory allocated with scsi_malloc() is indeed not correct.

To me the memory leaks reported by Valgrind seem unrelated to the patches in this pull request. Adding `free(initiatorname1); free(initiatorname2);` near the end of main() is probably sufficient to suppress these leak reports.

File: lib/scsi-lowlevel.c (hunk 3 above)
Comment (Ser01x) [reply lib/scsi-lowlevel.c]: fixed

File: test-tool/test_receive_copy_results_copy_status.c (hunk 4 above)
Comment (Ser01x) [reply test-tool/test_receive_copy_results_copy_status.c]: fixed

**Pull request merged**